### ISSUE TITLE
image/gif corrupted in older viewers

### ISSUE THREAD

**Issue by tenox7** [golang/go#33748]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12.5 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
yes, any release

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/usr/local/google/home/tenox/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/usr/local/google/home/tenox/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/lib/google-golang"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/google-golang/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build197159774=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Create a simple gif image:

<pre>
package main

import (
	"image"
	"image/color"
	"image/gif"
	"os"
	"fmt"
)

func main() {
	g, _ := os.Create("go.gif")
	defer g.Close()
	p := []color.Color{color.White, color.Black}
	r := image.Rect(0,0,100,100)
	i := image.NewPaletted(r, p)
	for x:=20; x<40; x++ {
		for y:=20; y<40; y++ {
			fmt.Printf("%d %d", x, y)
			i.SetColorIndex(x,y, 1)
		}
	}
	gif.Encode(g, i, nil)
}
</pre>

### What did you expect to see?

Expected to open in older image viewers, editors and browsers.

### What did you see instead?

A number of old browsers, image viewers and editors report generated gif image as corrupted or display it mangled. For example xv image viewer, https://en.wikipedia.org/wiki/Xv_(software), OmniWeb browser or most graphical software on SGI IRIX.

**Comment by odeke-em** [comment]
Thank you for reporting this issue @tenox7!

Kindly pinging some folks who might be able to help @nigeltao @horgh.

**Comment by smasher164** [comment]
Xv states that the file is truncated.
![xverr](https://user-images.githubusercontent.com/12636891/63414011-f4d53600-c3c8-11e9-84a0-7a33ad061d9a.png)

I assume you mean <= Omniweb 5, since Omniweb 6 renders the gif properly.
![correctomniweb6](https://user-images.githubusercontent.com/12636891/63414132-2f3ed300-c3c9-11e9-9acd-68a2f7daa95c.png)

**Comment by tenox7** [comment]
Yes OmniWeb 2.x and 3.x. I have updated the description, thank you.

**Comment by tenox7** [comment]
XV debug mode says this:

```
LoadGIF: imagesep (got=0)    at start: offset=0x14
xv: LoadGIF() - picture is 100x100, 1 bits, non-interlaced
  at end:   dataptr=0xc9

LoadGIF: trailerxv: Desired colormap
(  0  ff,ff,ff 9978)
(  1  00,00,00 22)


xv: result of sorting colormap
(  0  ff,ff,ff)     (  1  00,00,00)     
```

**Comment by nigeltao** [comment]
It's possibly the same issue as https://github.com/golang/go/issues/13746#issuecomment-215315656 with the eog ("Eye of GNOME") image viewer.

It's hard to say definitively, though. It's fair enough to say that Go's GIF encoder is writing something that old image viewers can't read, but it's hard to do something about that (in the Go code) without knowing what it is specifically those old image viewers don't like. And figuring out what they don't like probably means spelunking in xv or OmniWeb code, not in Go code.

Or, as I said in issue #13746, if it's:

> For example, the "GIF image loader cannot understand this image" error message comes from https://github.com/GNOME/gdk-pixbuf/blob/master/gdk-pixbuf/io-gif.c#L634 in the lzw_read_byte function, and the comment immediately above that says: "FIXME - we should handle this case".

then it's really up to EoG to fix their bug. Which they might already have done (in https://github.com/GNOME/gdk-pixbuf/commit/b88f1ce91a610a4e491a4ad6352183791e78afac, not sure if it's rolled out to stable distro's yet).

**Comment by tenox7** [comment]
I also think that this may be to do with LZW compression. Would it be possible to add an option to disable LZW compression when creating GIFs?

**Comment by nigeltao** [comment]
I don't think you can have a GIF image without LZW compression.

**Comment by mntn-xyz** [comment]
The author of the stb library, which also has trouble parsing these images, says the problem may be with the lack of a "clear code" which should be present per the GIF RFC: https://github.com/nothings/stb/issues/1222#issuecomment-938027720

**Comment by mntn-xyz** [comment]
OK, so with some digging into the actual bytes I am certain that the issue has something to do with the format of the data bytes.

Original file:
```
Header
  47 49 46 38 39 61

LSD
  0100 width (little endian)
  0100 height (little endian)
  80 packed field
    1 global color table flag
    000 color resolution
    0 sort flag
    000 size of global color table
  00 background color index
  00 aspect ratio

Global color table
  000000 color 1
  000000 color 2

Image descriptor
  2c separator
  0000 x location
  0000 y location
  0100 width
  0100 height
  00 packed field
    0 local color table flag
    0 interlace flag
    0 sort flag
    00 reserved
    000 size of local color table

Image data
  02 LZW minimum code size
  01 number of bytes
  28 data
  00 block terminator

Trailer
  3b
```

Fixed file:
```
Header
  47 49 46 38 39 61

LSD
  0100 width (little endian)
  0100 height (little endian)
  f0 packed field
    1 global color table flag
    111 color resolution
    0 sort flag
    000 size of global color table
  00 background color index
  00 aspect ratio

Global color table
  000000 color 1
  000000 color 2

Application extension block
  21 GIF extension code
  f9 application extension label (graphic control label)
  04 block size
  00 packed field
    000 reserved
    000 disposal method (not animated)
    0 user input flag
    0 transparency flag
  0000 delay time
  00 transparency index
  00 block terminator

Image descriptor
  2c separator
  0000 x location
  0000 y location
  0100 width
  0100 height
  00 packed field
    0 local color table flag
    0 interlace flag
    0 sort flag
    00 reserved
    000 size of local color table

Image data
  02 LZW minimum code size
  02 number of bytes
  4401 data
  00 block terminator

Trailer
  3b
```

Disregarding the application extension block (GraphicsMagick must have felt this was necessary despite the lack of animation), it's obvious that there is a difference in the image data. I manually edited the hex data of the broken image to match the data of the fixed image, and lo and behold, the image is now working. Here's the hex of the manually corrected image:

```
00000000: 4749 4638 3961 0100 0100 8000 0000 0000  GIF89a..........
00000010: 0000 002c 0000 0000 0100 0100 0002 0244  ...,...........D
00000020: 0100 0a                                  ...
```

The new image is one byte longer.

**Comment by mntn-xyz** [comment]
OK, checking the LZW bytes, it appears that the author of the stb library was correct. The code is not outputting a clear code before it begins writing the LZW data. Every source I come across treats this as mandatory, except for the W3C spec which says it "should" be included.

Perhaps the blockWriter setup() function could write a clear code (here: https://cs.opensource.google/go/go/+/refs/tags/go1.16.8:src/image/gif/writer.go;l=73;drc=refs%2Ftags%2Fgo1.16.8)? The clear code is 2 * the LZW code size (litWidth as defined here: https://cs.opensource.google/go/go/+/refs/tags/go1.16.8:src/image/gif/writer.go;l=327;drc=refs%2Ftags%2Fgo1.16.8;bpv=0;bpt=1).

**Comment by nigeltao** [comment]
@mntn-xyz thanks for the excellent and thorough debugging. I have sent out a fix for review.

**Comment by tenox7** [comment]
Thank you for fixing this!!!

**Comment by makeworld-the-better-one** [comment]
Great to see this fixed, thank you! I assume the fix will find its way into the next release, Go 1.17.3?
